library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -1223,19 +1223,29 @@ impl<T, A: Allocator> VecDeque<T, A> {
 
     /// Given a range into the logical buffer of the deque, this function
     /// return two ranges into the physical buffer that correspond to
-    /// the given range.
-    fn slice_ranges<R>(&self, range: R) -> (Range<usize>, Range<usize>)
+    /// the given range. The `len` parameter should usually just be `self.len`;
+    /// the reason it's passed explicitly is that if the deque is wrapped in
+    /// a `Drain`, then `self.len` is not actually the length of the deque.
+    ///
+    /// # Safety
+    ///
+    /// This function is always safe to call. For the resulting ranges to be valid
+    /// ranges into the physical buffer, the caller must ensure that for all possible
+    /// values of `range` and `len`, the result of calling `slice::range(range, ..len)`
+    /// represents a valid range into the logical buffer, and that all elements
+    /// in that range are initialized.
+    fn slice_ranges<R>(&self, range: R, len: usize) -> (Range<usize>, Range<usize>)
```

Comment:
Maybe it would be a bit clearer if this were a free function that takes `range, capacity, len` as arguments instead of the awkward mix where it takes the capacity from `self` but the `len` is fed in externally.

Reply:
I'm not sure that'd actually simplify things. It would always be incorrect to call `slice_ranges` with a capacity that's not just `self.capacity()`, whereas `len` doesn't have the same restriction. Making it a free function would also make both the implementation and the call sites more lengths and probably less readable (e.g. it couldn't use `self.to_physical_idx`) and might introduce more chances to accidentally mix up two arguments of the same type.